### DISCUSSION TITLE
WiringFactory is checked before explicitly registered behavior

### DISCUSSION THREAD

**Discussion by craig-day** [graphql-java/graphql-java#3163] [General]
**Describe the bug**
During the schema generation, building the code registry checks the `WiringFactory` to see if it can provide an item before checking explicitly registered items. This seems counter-intuitive. I would expect explicit registration of a data fetcher, type resolver, scalars, etc... to have the highest priority.

I think this could be as simple as reversing the if/else checks here to look for an explicit one first, then factory:

Scalars: https://github.com/graphql-java/graphql-java/blob/v20.1/src/main/java/graphql/schema/idl/SchemaGeneratorHelper.java#L438-L443

Interface type resolvers: https://github.com/graphql-java/graphql-java/blob/v20.1/src/main/java/graphql/schema/idl/SchemaGeneratorHelper.java#L504-L514

Union type resolvers: https://github.com/graphql-java/graphql-java/blob/v20.1/src/main/java/graphql/schema/idl/SchemaGeneratorHelper.java#L526-L536

Data Fetchers: https://github.com/graphql-java/graphql-java/blob/v20.1/src/main/java/graphql/schema/idl/SchemaGeneratorHelper.java#L836-L841

**To Reproduce**

Create wiring and register both an explicit type resolver in addition to a wiring factory. 

```java
var factory = new WiringFactory() {
  @Override
  public boolean providesTypeResolver(InterfaceWiringEnvironment environment) {
    return true;
  }

  @Override
  public TypeResolver getTypeResolver(InterfaceWiringEnvironment environment) {
    return env -> env.getSchema().getQueryType();
  }
};

var wiring = RuntimeWiring.newRuntimeWiring()
    .wiringFactory(factory)
    .type(TypeRuntimeWiring.newTypeWiring("MyType").typeResolver(env -> "FooType"))
    .build();
```

When we query for something that returns `MyType`, we will receive an error about failing to resolve `MyType` to a concrete implementation.

**Comment by bbakerman** [comment]
> building the code registry checks the WiringFactory to see if it can provide an item before checking explicitly registered items. This seems counter-intuitive. I would expect explicit registration of a data fetcher, type resolver, scalars, etc... to have the highest priority.

You have this described as bug but its a design choice.  We chose deliberately that the Wiring factory is the highest precedence when "wiring" code in.

This is why we have a two stage `providesXXX` and `getXXX` set of methods

(if we did this again we would probably have 1 method returning Optional but this choice was back in Java 6)

We ask the WiringFactory - do you want to wire in code here and if it says true we call it - we even assert you must

```
        if (wiringFactory.providesTypeResolver(environment)) {
            typeResolver = wiringFactory.getTypeResolver(environment);
            assertNotNull(typeResolver, () -> "The WiringFactory indicated it provides a interface type resolver but then returned null");

```

Our choice is that the "code driven" callback trumps the next levels.

Can you expand your intuition on how this should work and maybe some more background on how you code gets assembled and hence how this might be problematic for you

**Comment by craig-day** [comment]
Sorry, I know this isn't really a bug but I wasn't sure what else to qualify it as since I didn't the decision documented anywhere. Did I miss the docs, or could we explicitly document this in the wiring docs for the factory? 

I generally assumed explicit definitions should always be favored over dynamic lookup or programatic overrides. When my app boots, we use a CDI extension to discovery data fetchers and type resolvers and add them to our runtime wiring. We also provide a wiring factory which will provide a data fetcher that does generic to-string type lookups. I understand the design decision and can adjust my expectation. 

I do have a question on testing though. I feel like it makes the echoing wiring factory (or something similar) harder to use. I would like to load my schema and then wire up the one or two fetchers and loaders I'm testing, then would offload the rest of the stubbing to the factory. It seems like the best option here is to implement the `provides*` methods to check for explicit registration rather than returning `true` if I want the factory to be treated as a fallback/default?

**Comment by bbakerman** [comment]
> When my app boots, we use a CDI extension to discovery data fetchers and type resolvers and add them to our runtime wiring. 

Given our design decision to put WiringFactories first, why not change this so that the CDI discovery puts these type resolvers into a map inside a wiring factory and then the `providesTypeResolver` would check for presence in that map?

You already have a automatic mechanism for discovery - just change the system of registry

**Comment by bbakerman** [comment]
I have improved the javadoc around this
 
 https://github.com/graphql-java/graphql-java/pull/3162
 
 I am going to move this to a discussion - since this is a design choice - not a bug
